### ISSUE TITLE
Strange RGB Monitor Memory Corruption

### ISSUE THREAD

**Issue by audetto** [AppleWin/AppleWin#1018]
Hi

While I was working with Oliver Schmidt on porting the Uthernet II code to Windows, we hit this strange problem

![image](https://user-images.githubusercontent.com/212523/148773533-10742b9f-9507-4712-a264-bae43fed3c03.png)

What is strange is that it only happens with the RGB Card, none of the other video modes.

Do you believe it is simply a Apple ][ memory corruption (caused by the software that run before) or is it a AppleWin RGB renderer bug (since it only happens with RGB)?

[test.zip](https://github.com/AppleWin/AppleWin/files/7839105/test.zip)

**Comment by tomcw** [comment]
It's not a bug (as far as I understand), but a feature of the RGB card.
In TEXT mode, main memory ($400-7FF) provides the regular TEXT screen as always, but aux memory ($400-7FF) provides the foreground & background colours.

**Comment by inexorabletash** [comment]
Per @apple2geek the RGB cards should only producing colored text if AN3 is on. Does anything in the test leave that set?

**Comment by apple2geek** [comment]
Quick clarification - it's the other way around.  You get colored text if AN3 is OFF, and black & white if it's ON.
A RESET cycle sets all the annunciators to OFF in hardware, but the standard Apple ROM reset routine explicitly sets AN2 and AN3 back to ON, making the text un-colored.  Some alternate ROMs don't bother to tweak the annunciators, and can leave you with wacky colored text.

**Comment by tomcw** [comment]
> You get colored text if AN3 is OFF, and black & white if it's ON.

From the save-state (test.zip, above):
```
Annunciator3: false
```

**Comment by apple2geek** [comment]
So if AN3 is OFF (presumably what "false" means), then the displayed behavior - colored text - is correct.
But that's not the default setting for AN3 either following a RESET or a power-on, so if AN3 is indeed OFF, something has to have explicitly set it that way.

**Comment by audetto** [comment]
What about this:

https://github.com/AppleWin/AppleWin/blob/a727db7cc509cbd897942be24e76247b9d1c6ff4/source/Memory.cpp#L2146

Nothing else touches the code when I run the offending software. It starts as 0 = OFF.

**Comment by apple2geek** [comment]
That may be how the code for the emulator sets the annunciator states initially, and that's fine.  The _default_ state for all four annunciators on a physical hardware Apple II on power-up, or when a hardware RESET occurs _is_ (at least initially) to have all four of them set to "OFF" by the IOU.

**BUT** the Apple System ROM reset-handler routine (starting at $FA62 on the //e) which takes over immediately _explicitly_ resets the annunciators back to different states (OFF/OFF/ON/ON) at $FA6F:
```
FA62: D8           RESET           CLD                   ;Do this 1st this time;
FA63: 20 84 FE                     JSR   SETNORM         ;Set Normal Video Mode & ...
FA66: 20 2F FB                     JSR   INIT            ;Init Keyboard & Screen as I/O Devices
FA69: 20 93 FE                     JSR   SETVID          ;Reset CSW Output Vector to COUT1 <--+/
FA6C: 20 89 FE                     JSR   SETKBD          ;Reset KSW Input Vector to KEYIN <--+/

FA6F: AD 58 C0     INITAN          LDA   AN0OFF          ; Turn OFF: AN0
FA72: AD 5A C0                     LDA   AN1OFF          ; Turn OFF: AN1
FA75: AD 5D C0                     LDA   AN2ON           ; Turn  ON: AN2
FA78: AD 5F C0                     LDA   AN3ON           ; Turn  ON: AN3

FA7B: AD FF CF                     LDA   CLRROM          ;Tell slots to disable their extension ROM
FA7E: 2C 10 C0                     BIT   KBDSTRB         ;Clear KBD Strobe (KBD<128)
...
...
FFFA: FB 03                        DW    NMI             ;NMI Vector
FFFC: 62 FA                        DW    RESET           ;RESET Vector
FFFE: 40 FA                        DW    IRQ             ;IRQ Vector
```
So, following a power-on, or a RESET, annunciators 0/1/2/3 should always end up OFF/OFF/ON/ON because the System ROM explicitly sets them that way.  If AN3 somehow ends up staying OFF following a reset or power-on, either some code other than the system ROM is toggling it back to OFF again, or else the system ROM's explicit toggle to ON at $FA78 is failing.  Or perhaps being ignored.
Is it possible the emulator isn't correctly handling all of the C058-C05F annunciator softswitches?

If AN3 is OFF, you should expect to get the colored text from the RGB card, like the screen-shot at the top.

**Comment by apple2geek** [comment]
I'm not the **sharpest** coder in the book, but it does look like the softswitches for the annunciators are being properly handled by the code in Memory.cpp, with the relevant code that actually changes the annunciator's state being line 731:

`731	g_Annunciator[(address>>1) & 3] = (address&1) ? true : false;`

so the access to $C05F at $FA78 *should* be setting g_Annunciator[3] to "true" (aka ON) unless I'm missing something?

If the state of the annunciator is being toggled and tracked correctly, then perhaps the code that generates the RGB display is either not checking the AN3 state, or is getting its check backwards?

**Comment by audetto** [comment]
But

https://github.com/AppleWin/AppleWin/blob/a727db7cc509cbd897942be24e76247b9d1c6ff4/source/Memory.cpp#L559-L564

I never see the debugger hitting `IO_Annunciator` for `C05F`

**Comment by apple2geek** [comment]
Ok, wait - I think I see at least part of the problem.  The softswitches C05E and C05F are being handled differently than all the other annunciator softswitches.  At lines 530-535 and 559-564, the AN3 softswitches are being used to call GetVideo().VideoSetMode() via the return() if the system is not a ][+ (so a //e or a //c) and SW_IOUDIS is false.  But this means that the STATE of AN3 isn't being tracked in this case, only the video mode is being modified.

This logic needs to be changed.  Annunciator 3 is still a legitimate line, even on the //e and //c, and its state needs to be tracked, so the call to GetVideo().VideoSetMode() still needs to also call IO_Annunciator().

**Comment by apple2geek** [comment]
How does the RGB emulation decide whether to display colored text or black & white?  Does it explicitly read the state of Annunciator 3, or does the emulation expect GetVideo().VideoSetMode() to set the right mode for the RGB?

If GetVideo().VideoSetMode() isn't setting the RGB mode to colored/non-colored text, that's also a problem.

**Comment by apple2geek** [comment]
I see where the RGB colored/b&w text "mode" is being handled, towards the end of RGBMonitor.cpp.
There are short routines to change and query the mode flag (g_nTextFBMode) from lines 1361-1374.
RGB_EnableTextFB(), RGB_DisableTextFB(), RGB_IsTextFB()

Where are RGB_EnableTextFB() and RGB_DisableTextFB() called from?  I don't see any reference to them in the Video::VideoSetMode, or anywhere in Video.cpp AT ALL, for that matter.

What toggles g_nTextFBMode?  It should follow AN3, same as HGR/DHGR.

**Comment by tomcw** [comment]
It looks like there's definitely an emulator issue, as AN3 isn't being checked here:
https://github.com/AppleWin/AppleWin/blob/5fbd8339c10e3dd623d9613a47986ff95658fe6b/source/NTSC.cpp#L1978-L1986

**Comment by apple2geek** [comment]
So it looks like RGB_EnableTextFB() and RGB_DisableTextFB() are called from inside NTSC_SetVideoMode() which is itself called from inside NTSC_VideoUpdateCycles().  I'm not sure how it's called beyond that.

The logic for both Annunciator 3 (g_Annunciator[3]) and the colored text flag (g_nTextFBMode) needs to be cleaned up to match reality.
1 - The enabled/disabled state of AN3 needs to be tracked and saved as enabled/disabled, regardless of machine type, so don't skip the IO_Annunciator() call during C05E/C05F softswitch handling
and
2 - the colored text flag needs to follow the state of Annunciator 3, so when g_Annunciator[3] = 0, g_nTextFBMode = 1 and vice-versa.

**Comment by tomcw** [comment]
Hmm... digging a bit more, and again the emulation looks fine:

For //e, AN3 gets tracked in `g_uVideoMode` in bitmask `VF_DHIRES` (0x02).

So r/w accesses to $C05E/F (for a //e or Enhanced //e) will not set/clear AN3 (`g_Annunciator[3]`), but instead will call `GetVideo().VideoSetMode()`, which sets/clears `VF_DHIRES` in `g_uVideoMode`.

From the save-state:
```
      Video Mode: 0x00000042
```
Which decodes to: VF_TEXT(0x40) | VF_DHIRES(0x02).

NB. Editing the save-state to `Video Mode: 0x00000040` now does _not_ render TEXT in this multi-coloured RGB mode.

So something previously has accessed $C05E, turning off AN3 and therefore setting DHIRES video mode.

**Comment by apple2geek** [comment]
> So r/w accesses to $C05E/F (for a //e or Enhanced //e) will not set/clear AN3 (g_Annunciator[3]), but instead will call GetVideo().VideoSetMode(), which sets/clears VF_DHIRES in g_uVideoMode.

This is not correct, and does not match the behavior of the physical hardware.  $C05E/F do in fact still toggle AN3 physically on a //e, and should be emulated as such.  The AN3 line is present on both the //e's internal game port and it AUX slot.

If the GetVideo().VideoSetMode() call is going to be used to handle emulating AN3's operation of the AUX RGB card, that's fine, but then the call to GetVideo().VideoSetMode() must update not just the DHGR mode, but the FB text flag as well.  The SAME signal line - AN3 - is used to control both DHGR and the colored 40-column text on the physical RGB card hardware.

No matter what, you still need to call IO_Annunciator() for $C05E/F access.

**Comment by tomcw** [comment]
>$C05E/F do in fact still toggle AN3 physically on a //e, and should be emulated as such. The AN3 line is present on both the //e's internal game port and its AUX slot.

@apple2geek - thanks for confirming that point about AN3 still existing on the //e.

**Comment by tomcw** [comment]
btw, whilst on the subject of AN3/DHIRES...
Is $C07F _really_ supported by the Enhanced //e? Or just //c (and above)?

In the Apple IIe Technical Manual(*), Table 2-10 "Display Soft Switches" describes IOUDIS, RDIOUDIS and RDDHIRES:

![AIIeTRM-Table2-10](https://user-images.githubusercontent.com/6696896/149021200-32514486-a7a0-4a22-a054-3a9b10cef761.PNG)

I'm guessing this was accidentally pasted in from the //c TRM.

(*) (C) 1985
ISBN 0-201-17720-X
First printing July 1985

**Comment by apple2geek** [comment]
I don't believe the //e supported the IOUDIS stuff - I thought that was only the //c.  The naming of the signals suggests that's the case, too - the //c had no access to AN3 (it had no internal game port and no AUX slot), so the "AN3" designation was dropped, and it was explicitly labeled "DHIRES".  I can cross-check with a physical //e here if I can find some time.

**Comment by tomcw** [comment]
If you have an Enhanced //e, then can you try this (copied from this 2019 [issue](https://github.com/AppleWin/AppleWin/issues/636#issuecomment-480530115)):

Test-1:
- `RESET`
- PR#3
- CALL -151
- C050 N C053 N C057 ; GR + MIXED + HIRES
- C07F ; returns eg. 30 or A0
- C05E ; DHIRES on
- C07F ; returns eg. 30 or A0

Test-2:
- `RESET`
- PR#3
- CALL -151
- C050 N C053 N C057 ; GR + MIXED + HIRES
- C07E:0
- C07E ; returns eg. 30 or A0
- C07F:0
- C07E ; returns eg. 30 or A0

NB. for the lines that say "returns eg. 30 or A0": just repeat these a number of times. Hopefully you'll get values with both bit7 set & clear, and thus showing that the Enhanced //e doesn't support these readable switches.

**Comment by tomcw** [comment]
OK, AN3 is now supported for the //e models (see af81434).

But @audetto this makes no difference to the multi-coloured TEXT mode you reported at the start of this issue.
Basically, some code is setting AN3 off via an access to $C05E (ie. enabling DHIRES) and for an RGB video card you will still see this.

**Comment by audetto** [comment]
There is indeed another access to $C05E which flips AN3 (which is not tracked correctly).

The multi-color text is still there, but if I access C05E or C05F I can enable and disable it.
Everything is behaving correctly I believe.

**Comment by inexorabletash** [comment]
FWIW, I believe MAME's RGB card emulation currently has a similar bug where flipping AN3 is not sufficient to exit multi-color text mode.

**Comment by tomcw** [comment]
@inexorabletash - just to be clear, the only bug in AppleWin was that AN3 wasn't being set/cleared for the //e models... _but_ on those //e models, the DHIRES video mode was being set correctly. So since AN3 itself isn't used by anything (as AppleWin doesn't support the Annunciator outputs), then this bug wasn't actually affecting any behaviour.

**Comment by tomcw** [comment]
Tagging this issue with the 'bug' label, since there were some bugs (af81434), but the original reported behaviour (multi-coloured text for an RGB card when AN3 is off / DHIRES is enabled) is spec.

**Comment by tomcw** [comment]
Closing this as the AN3 bug was fixed in 1.30.8.0.
And the other point: "Is $C07F really supported by the Enhanced //e? Or just //c (and above)?" is covered in #636.
